### PR TITLE
feat: stash unstaged changes with lint-staged

### PR DESCRIPTION
**Pull request by varl**
This works around the problems in #84 and #85 by using `lint-staged` from NodeJS instead of as an executable, so it works when the tool is used globally.

This also enables the functionality which #72 adds, but using `lint-staged` sophisticated tree diffing when applying and staging hunks, which works around some of the problems with the implementation in #72.

BREAKING CHANGE: Changes multiple defaults for validate, and apply

- `d2-style js apply` no longer stages files by default, use `--stage` to stage fixed files. The intended usage for this function is primarily to apply the js code standards to files in general, e.g. unstaged/staged files or an entire repo.

- `d2-style validate` has changed to by default run `--fix`, and uses `lint-staged` to only apply code standards to staged files and/or hunks.

- `d2-style validate --all` no longer exists, `validate` only operates on staged files and is intended to be used as a pre-commit check. Use `d2-style js check` or `apply` with `--all` instead.

- `d2-style validate` output has changed. It no longer gives detailed information during a run. It only prints errors when it fails.

- `d2-style js install` has been removed, and the clean-up functions have also been removed.

### CODE REVIEW

**Comment by varl** [comment]
Configuration can be overridden:

```bash
d2-style validate js --prettier-config .prettierrc.js
```

```bash
d2-style validate js --eslint-config .eslintrc.js
```

**Comment by varl** [comment]
The `tools/cli.js` file is a wrapper for the tool runners, to allow `lint-staged` to execute the same runners as an external script.

File: src/cmds/validate.js [hunk 1]
```diff
@@ -20,82 +18,63 @@ exports.builder = {
...
+    }
+
+    lintStaged({
+        configPath: LINT_STAGED_CONFIG,
+        quiet: true,
+        debug: false,
+    })
+        .then(s => {
+            if (!s) {
+                process.exit(1)
```
[the remark this answers is not in the document]
Comment (varl) [reply src/cmds/validate.js]: The error is printed else-where, so you get something like this:

```bash
✖ node /home/varl/dev/dhis2/libs/cli-style/src/tools/cli.js js false /home/varl/dev/dhis2/apps/usage-analytics/.commitlintrc.js,/home/varl/dev/dhis2/apps/usage-analytics/.dependabot/config.yml,/home/varl/dev/dhis2/apps/usage-analytics/.eslintrc.js,/home/varl/dev/dhis2/apps/usage-analytics/.github/semantic.yml found some errors. Please fix them and try committing again.

[ERROR] 1 file(s) violate the code standard.
[ERROR] /home/varl/dev/dhis2/apps/usage-analytics/.commitlintrc.js
[ERROR] Not formatted according to standards.
```

File: src/cmds/validate.js [hunk 2]
```diff
@@ -20,82 +18,63 @@ exports.builder = {
...
-        if (report.hasViolations) {
-            violations += report.violations.length
-        }
-
-        if (fix) {
-            const fixed = report.fix()
-            fixedFiles.push(...fixed)
-        }
-    }
+    const { fix, group = ['all'], stage, eslintConfig, prettierConfig } = argv
```
[the remark this answers is not in the document]
Comment (varl) [reply src/cmds/validate.js]: It's been like this for a while now, `group` is tied to the command line variable name:

```
  d2 style setup [group..]     Setup DHIS2 configurations for a/all group(s)
  d2 style validate [group..]  Validate DHIS2 configurations for a/all group(s)
```

File: src/groups.js [hunk 3]
```diff
@@ -14,49 +25,25 @@ const groups = {
...
+            [ESLINT_CONFIG, path.join('.eslintrc.js')],
+            [PRETTIER_CONFIG, path.join('.prettierrc.js')],
+            [BROWSERSLIST_CONFIG, path.join('.browserslistrc')],
+            [COMMITLINT_CONFIG, path.join('.commitlintrc.js')],
         ],
     },
 
     package: {
-        tools: [tools.package],
+        tools: [`${tools.package}`],
```
[the remark this answers is not in the document]
Comment (varl) [reply src/groups.js]: No, fixed in c6d9f09.

**Comment by varl** [comment]
> When having partially staged files, a `.rej` file will be created and **not** cleaned up afterwards
> 
> ```
> Untracked files:
>   (use "git add <file>..." to include in what will be committed)
> 
>         src/foo.js.rej
> ```
> 
> These should def. be removed

Seems like a bug in `lint-staged`: https://github.com/okonet/lint-staged/blob/e332c806fbe6912aa2e7c89b87295f31f2af29db/src/gitWorkflow.js#L133-L149

They actually try their best to clean up those `.rej` files. If they cannot remove them something strange has happened. 🤔

I've only seen it a couple of times, most of the times the `.rej` files are deleted correctly.

**Comment by varl** [comment]
> But I'd expect to get feedback on which files have changed..
> 
> When changing `debug` to `true` and `quiet` to `true`, the output contains the touched files, but not which ones have been modified, so I'm not sure whether this is possible or not..

The lack of control over the output is the part I like the least about introducing this PR.

1. `tools/cli.js` could figure out what files it is running on (the list is given to it from `lint-staged`).

2. `lint-staged` asynchronously prints log messages, so when printing manually I wound up with inter-twined log statements with no logical order which was worse than not having logs at all.

3. `lint-staged` breaks the way the `reporter` from `@dhis2/cli-helpers-engine` works, so those messages are eaten up.

4. The non-quiet formatter for `lint-staged` is shit.

Etc...

Some of these can be fixed and/or worked around.

**Comment by Mohammer5** [comment]
> Seems like a bug in lint-staged

Hmm this is bad.. If we don't work around this issue somehow (delete im in pre-push?), we'll endup with some `.rej` files in our repos.. 

> The lack of control over the output is the part I like the least about introducing this PR.

Hmm this is bad as well but not as bad as the one above as we won't care about what changed most of the time, so I could live with this for the time being

**Comment by varl** [comment]
> Hmm this is bad.. If we don't work around this issue somehow (delete im in pre-push?), we'll endup with some `.rej` files in our repos..

We can run a cleanup after `lint-staged` has done its business: abe23b5

**Pull request merged**